harrystech/tap-gladly
# DENG-2040: Implement TAP API conversions Gladly streams

## Definition

This PR implements the export file conversations by content type into a different stream since even though they're coming from the same API call, the schema is different by content type.

* **ExportJobsStream** : maps out to Export/List Jobs.
* **ExportFileConversationItemsChatMessage**: maps out to Export/Get file where content type in "CHAT_MESSAGE".
* **ExportFileConversationItemsConversationNote**: maps out to Export/Get file where content type in "CONVERSATION_NOTE".
* **ExportFileConversationItemsTopicChange**: maps out to Export/Get file where content type in "TOPIC_CHANGE".
* **ExportFileConversationItemsSms**: maps out to Export/Get file where content type in "SMS".
* **ExportFileConversationItemsConversationStatusChange**: maps out to Export/Get file where content type in "CONVERSATION_SATUS_CHANGE".
* **ExportFileConversationItemsPhoneCall**: maps out to Export/Get file where content type in "PHONE_CALL".
* **ExportFileConversationItemsVoiceMail**: maps out to Export/Get file where content type in "CHAT_MESSAGE".

## How to test

I have been testing by setting the environment variable that are picked automatically by the tap when they follow a certain format:

```
export TAP_GLADLY_START_DATE=2022-09-06T00:00:00Z
export TAP_GLADLY_PASSWORD=...
export TAP_GLADLY_USERNAME=...
export TAP_GLADLY_API_URL_BASE=...
```
Follow the steps in the Readme to setup the sdk development environment, then you can test by running:

```shell
poetry run tap-gladly --config=ENV > output.json
cat output.json
```

You will be able to see the different streams:

```
cat output.json  | jq '. | select( .stream=="jobs") '
cat output.json  | jq '. | select( .stream=="conversation_chat_message")'
cat output.json  | jq '. | select( .stream=="conversation_conversation_note")'
cat output.json  | jq '. | select( .stream=="conversation_topic_change")'
cat output.json  | jq '. | select( .stream=="conversation_sms")'
cat output.json  | jq '. | select( .stream=="conversation_conversation_status_change")'
cat output.json  | jq '. | select( .stream=="conversation_phone_call")'
cat output.json  | jq '. | select( .stream=="conversation_voicemail")'
```

Each of these stream will be picked up by a loader and validated against the json schema in `tap_gladly/schemas`.

### Test with meltano

This was tested alongside the `target_jsonl` loader:

```
meltano elt tap-gladly target-jsonl
```

using the config (providing settings using environment variables, the same as above):
```YAML
version: 1
default_environment: dev
environments:
- name: dev
- name: staging
- name: prod
project_id: harrys-data-warehouse-d2d8f490-c200-4863-9f11-d0b25a259426
plugins:
  extractors:
  - name: "tap-gladly"
    namespace: "tap_gladly"
    pip_url: git+https://github.com/harrystech/tap-gladly.git@ynaim94/deng-2040/tap-api-gladly
    capabilities:
    - state
    - catalog
    - discover
    settings:
    - name: username
    - name: password
      kind: password
    - name: api_url_base
    - name: start_date
  loaders:
  - name: target-jsonl
    variant: andyh1203
    pip_url: target-jsonl
```

```
ls -l output/
total 4016
-rw-r--r-- 1 root root  236977 Sep 20 22:44 conversation_chat_message.jsonl
-rw-r--r-- 1 root root   49349 Sep 20 22:44 conversation_conversation_note.jsonl
-rw-r--r-- 1 root root  871653 Sep 20 22:44 conversation_conversation_status_change.jsonl
-rw-r--r-- 1 root root  439911 Sep 20 22:44 conversation_phone_call.jsonl
-rw-r--r-- 1 root root  232002 Sep 20 22:44 conversation_sms.jsonl
-rw-r--r-- 1 root root 2139325 Sep 20 22:44 conversation_topic_change.jsonl
-rw-r--r-- 1 root root    2209 Sep 20 22:44 conversation_voicemail.jsonl
-rw-r--r-- 1 root root    1184 Sep 20 22:44 jobs.jsonl
```
